File: 4week/3task/Car/Car.java
public class Car {
    private double fuelAmount;
    private double fuelCapacity;
    private double fuelConsumption;
    private Point location;
    private String model;

    public Car(double capacity, double consumption, Point location, String model) {
        if ( capacity < 0 ) {
            capacity = 60;
        }
        if ( consumption < 0 ) {
            consumption = 0.6;
        }
        this.fuelAmount = capacity;
        this.fuelCapacity = capacity;
        this.fuelConsumption = consumption;
        this.location = location;
        this.model = model;
    }

    public Car(double capacity, double consumption, double x, double y, String model) {
        Point location = new Point(x, y);
        if ( capacity < 0 ) {
            capacity = 60;
        }
        if ( consumption < 0 ) {
            consumption = 0.6;
        }
        this.fuelAmount = capacity;
        this.fuelCapacity = capacity;
        this.fuelConsumption = consumption;
        this.location = location;
        this.model = model;
    }

    public Car(double capacity, double consumption, String model) {
        this(capacity, consumption, 0, 0, model);
    }

    public Car() {
        this(60, 0.6, 0, 0, "Mercedes");
    }

    public double getFuelAmount() {
        return this.fuelAmount;
    }

    public double getFuelCapacity() {
        return this.fuelCapacity;
    }

    public double getFuelConsumption() {
        return this.fuelConsumption;
    }
    
    public Point getLocation() {
        return this.location;
    }

    public String getModel() {
        return this.model;
    }

    public void drive(Point destination) throws OutOfFuelException {
        double fuelNeeded;
        double distance;

        distance = location.distance(destination);
        fuelNeeded = distance * fuelConsumption;
        if ( fuelNeeded > fuelAmount ) {
            throw new OutOfFuelException();
        }
        fuelAmount -= fuelNeeded;
        location = destination;
    }

    public void drive(double x, double y) throws OutOfFuelException {
        drive(new Point(x,y));
    }

    public void refill(double fuel) throws ToMuchFuelException {
        double NewFuelAmount = fuelAmount;
        NewFuelAmount += fuel;
        if (NewFuelAmount > fuelCapacity) {
            throw new ToMuchFuelException();
        }
        fuelAmount = NewFuelAmount;
    }
    
    public void print() {
        System.out.println("Model: "+ this.model + "\n"
                            +"Location: " + this.location + "\n"
                            +"Fuel consumption: " + this.fuelConsumption + "\n"
                            +"Fuel amount: " + this.fuelAmount + "\n"
                            +"Fuel capacity: " + this.fuelCapacity);
    }
}

File: 4week/1task/Car/Car.cpp
#include <iostream>
#include "Point.h"
#include "Car.h"

Car::Car(double capacity, double consumption, const Point& location, const std::string& model) {
    this->fuelAmount = capacity;
    this->fuelCapacity = capacity;
    this->fuelConsumption = consumption;
    this->location = location;
    this->model = model;
}

Car::~Car() {}

double Car::getFuelAmount() const {
    return this->fuelAmount;
}

double Car::getFuelCapacity() const {
    return this->fuelCapacity;
}

double Car::getFuelConsumption() const {
    return this->fuelConsumption;
}

const Point& Car::getLocation() const {
    return this->location;
}

const std::string& Car::getModel() const {
    return this->model;
}

void Car::drive(const Point& destination) {
    
    Point locationCar = getLocation();
    double distance = this->location.distance(destination);
    double fuelNeeded = distance * this->fuelConsumption;
    
    if (fuelNeeded > this->fuelAmount) {
        throw OutOfFuel();
    }
    this->location = destination;
    this->fuelAmount -= fuelNeeded;
}

void Car::drive(double x, double y) {
    Point locationAfterDrive(x, y);
    double distance = this->location.distance(locationAfterDrive);
    double fuelNeeded = distance * this->fuelConsumption;
    
    if (fuelNeeded > this->fuelAmount) {
        throw OutOfFuel();
    }
    this->location = locationAfterDrive;
    this->fuelAmount -= fuelNeeded;
}

void Car::refill(double fuel) {
        if ((fuel + this->fuelAmount) > this->fuelCapacity) {
            throw ToMuchFuel();
    }
    this->fuelAmount += fuel;
}

std::ostream& operator<<(std::ostream& out, const Car& car) {
    out << "Model: " << car.getModel() << "\n"
        << "Location: " << car.getLocation() << "\n"
        << "Fuel consumption: " << car.getFuelConsumption() << "\n"
        << "Fuel amount: " << car.getFuelAmount() << "\n"
        << "Fuel capacity: " << car.getFuelCapacity() << "\n";
    return out;
}
